File: cli/http.ts
// SPDX-License-Identifier: PMPL-1.0

import { SafeJson, type Result, err, ok } from "@proven/mod.ts";
import type { HttpOptions } from "./types.ts";

export type { Result };

/** Sleep for a given number of milliseconds */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * POST JSON to a service endpoint with retries and exponential backoff.
 */
export async function postJson(
  baseUrl: string,
  path: string,
  token: string | undefined,
  body: string,
  opts: HttpOptions,
): Promise<void> {
  const url = `${baseUrl.replace(/\/+$/, "")}${path}`;
  const attempts = opts.retries + 1;

  const headers: HeadersInit = {
    "Content-Type": "application/json",
  };
  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  let lastError: Error | null = null;

  for (let attempt = 0; attempt < attempts; attempt++) {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), opts.timeoutMs);

      const response = await fetch(url, {
        method: "POST",
        headers,
        body,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (response.ok) {
        return;
      }

      lastError = new Error(`HTTP error: ${response.status} ${response.statusText}`);

      if (attempt + 1 < attempts) {
        await sleep(opts.backoffMs * (attempt + 1));
      }
    } catch (err) {
      lastError = err instanceof Error ? err : new Error(String(err));

      if (attempt + 1 < attempts) {
        await sleep(opts.backoffMs * (attempt + 1));
      }
    }
  }

  throw lastError ?? new Error("Request failed");
}

/**
 * GET JSON from a service endpoint with retries and exponential backoff.
 * Uses proven's SafeJson for safe parsing.
 */
export async function getJson<T>(
  baseUrl: string,
  path: string,
  token: string | undefined,
  opts: HttpOptions,
): Promise<Result<T>> {
  const url = `${baseUrl.replace(/\/+$/, "")}${path}`;
  const attempts = opts.retries + 1;

  const headers: HeadersInit = {
    "Accept": "application/json",
  };
  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  let lastError = "Request failed";

  for (let attempt = 0; attempt < attempts; attempt++) {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), opts.timeoutMs);

      const response = await fetch(url, {
        method: "GET",
        headers,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (response.ok) {
        const text = await response.text();
        const parsed = SafeJson.parse(text);
        if (parsed.ok) {
          return ok(parsed.value as T);
        }
        return err(`JSON parse error: ${parsed.error}`);
      }

      lastError = `HTTP error: ${response.status} ${response.statusText}`;

      if (attempt + 1 < attempts) {
        await sleep(opts.backoffMs * (attempt + 1));
      }
    } catch (e) {
      lastError = e instanceof Error ? e.message : String(e);

      if (attempt + 1 < attempts) {
        await sleep(opts.backoffMs * (attempt + 1));
      }
    }
  }

  return err(lastError);
}

File: opsm_ex/lib/opsm/http.ex
# SPDX-License-Identifier: PMPL-1.0
defmodule Opsm.Http do
  @moduledoc """
  HTTP client with retries, exponential backoff, and connection pooling.
  Uses Req under the hood with Finch for connection management.

  All JSON responses are validated using Verified.Json for safety.

  Connection pooling (P5) improves performance by:
  - Reusing connections to the same host
  - Reducing TLS handshake overhead
  - Managing connection lifecycle automatically
  """

  alias Opsm.Types.HttpConfig
  alias Opsm.Verified.Json

  # Default pool configuration for CLI usage
  @default_pool_size 10

  @doc """
  Build a configured Req client with retries, timeout, and connection pooling.
  """
  def build_client(%HttpConfig{} = config, opts \\ []) do
    base_url = Keyword.get(opts, :base_url)
    token = Keyword.get(opts, :token)
    _pool_size = Keyword.get(opts, :pool_size, @default_pool_size)

    headers =
      if token do
        [{"authorization", "Bearer #{token}"}]
      else
        []
      end

    # Configure connection pooling (P5)
    # Req uses Finch under the hood - we configure pooling here
    Req.new(
      base_url: base_url,
      headers: headers,
      receive_timeout: config.timeout_ms,
      connect_options: [
        # Connection pool configuration
        timeout: config.timeout_ms,
        protocols: [:http1, :http2]
      ],
      retry: :transient,
      max_retries: config.retries,
      retry_delay: fn attempt ->
        # Exponential backoff
        config.backoff_ms * attempt
      end
    )
  end

  @doc """
  POST JSON to a path.
  """
  def post_json(client, path, body) do
    case Req.post(client, url: path, json: body) do
      {:ok, %Req.Response{status: status}} when status in 200..299 ->
        :ok

      {:ok, %Req.Response{status: status, body: body}} ->
        {:error, "HTTP error: #{status} - #{inspect(body)}"}

      {:error, reason} ->
        {:error, "Request failed: #{inspect(reason)}"}
    end
  end

  @doc """
  GET JSON from a path.

  Response is validated using Verified.Json for safety (depth/size limits).
  """
  def get_json(client, path) do
    case Req.get(client, url: path) do
      {:ok, %Req.Response{status: status, body: body}} when status in 200..299 ->
        # Validate JSON response if it's a string
        case body do
          body when is_binary(body) ->
            case Json.decode(body) do
              {:ok, parsed} -> {:ok, parsed}
              {:error, reason} -> {:error, "JSON validation failed: #{inspect(reason)}"}
            end

          body when is_map(body) or is_list(body) ->
            # Already parsed by Req, return as-is
            # Note: We trust Req's JSON decoder for now, but in v2.0
            # we should validate all JSON regardless of source
            {:ok, body}

          _other ->
            {:error, "Invalid response body type"}
        end

      {:ok, %Req.Response{status: status, body: body}} ->
        {:error, "HTTP error: #{status} - #{inspect(body)}"}

      {:error, reason} ->
        {:error, "Request failed: #{inspect(reason)}"}
    end
  end

  @doc """
  Build URL by joining base and path.
  """
  def build_url(base_url, path) do
    base = String.trim_trailing(base_url, "/")
    "#{base}#{path}"
  end

  @doc """
  Build and validate URL using Verified.Url.

  Returns `{:ok, url_string}` if valid, `{:error, reason}` otherwise.
  """
  def build_validated_url(base_url, path) do
    url_string = build_url(base_url, path)

    case Opsm.Verified.Url.validate(url_string) do
      {:ok, _validated} -> {:ok, url_string}
      {:error, reason} -> {:error, {:invalid_url, reason}}
    end
  end
end
